rust/main/agents/relayer/src/msg/metadata/message_builder.rs

```diff
@@ -0,0 +1,430 @@
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmCountReached(ISM_MAX_COUNT));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn max_depth_reached() {
+        let base_builder = build_mock_base_builder();
+
+        for _ in 0..3 {
+            insert_mock_routing_ism(&base_builder);
+        }
+
+        let ism_address = H256::zero();
+        let message = HyperlaneMessage::default();
+
+        let mut message_builder =
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+        message_builder.depth = ISM_MAX_DEPTH - 2;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmDepthExceeded(ISM_MAX_DEPTH));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn max_ism_count_reached() {
+        let base_builder = build_mock_base_builder();
+
+        for _ in 0..11 {
+            insert_mock_routing_ism(&base_builder);
+        }
+
+        let ism_address = H256::zero();
+        let message = HyperlaneMessage::default();
+
+        let message_builder =
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+
+        *message_builder.ism_count.lock().await = ISM_MAX_COUNT - 10;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmCountReached(ISM_MAX_COUNT));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn max_ism_count_reached_with_aggregate() {
+        let base_builder = build_mock_base_builder();
+
+        // push aggregation ISMs
+        //
+        // R = Routing (9)
+        // Agg = Aggregation
+        //
+        //           Agg
+        //     /      |      \
+        //    R       R       R
+        //    |       |       |
+        //    R       R       R
```

Comment:
for instance in this case I'd want us to check that the tracked height never exceeds 2 but that we do hit a depth of 2 three times